README.md

```diff
@@ -203,107 +203,78 @@ All messages must pass the following validations in addition to specific validat
+A cast without a `parent` is a top-level cast, which clients should display on the user's profile or timeline. A cast with a `parent` is a reply to another cast, web URL or on-chain object which should be displayed in a thread.
 
-1. `schema` must be known.
-2. `targetCastUri` must be a valid Farcaster Cast URI and must not reference this message.
+Casts form a series of trees where each root is a Cast or URI and each child node is a reply cast. Each tree can be rendered as a thread in the UI. Trees are guaranteed to be acyclic because a parent must be hashed and signed before a child can point to it. A parent that changes its parent to point to a child will change its hash and break its relationship with the child.
 
-#### Set Construction
+```mermaid
+graph TB 
+    A((1))-->B((2))
+    A-->C((3))
+    A-->D((4))
+    B-->E((5))
+    B-->F((6))
+    B-->G((7))
+    C-->H((8))
+```
 
-Recasts can be stored using a two-phase set just like Short Text Casts. The remove operation is identical, while the addition operation is performed with `CastRecast`, and has one additional rule:
+A Cast message must pass the following validation steps:
 
-1. If there is a CastRecast `a` in the add **add-set** with the same `targetCastUri` and `fid` as the incoming CastRecast `rc`:
-   - If `a > rc`, discard `d`
-   - If `a < rc`, delete `a` and add `rc` into the rem-set
+1. `text` must contain <= 320 valid unicode characters
+2. `embed` must contain between 0 and 2 `items`
+3. `item` must be a URI of at most 256 characters
+4. `parent`, if present, must be a valid URI not equal to this message's URI
 
-### 4.1.3 Remove Messages
+### 4.1.2 Remove Messages
 
-A remove instructs the set to remove a previously created cast. It is a type of soft delete where the content of the message is removed but its hash is retained forever. A user who has a copy of the remove message can prove that the message was posted at some point by the author.
+A _Cast Remove_ only contains a reference to the hash of the _Cast Add_. It allows for permanent deletion of casts while eliding the data of the original cast.
 
 ```ts
-type CastRemove = {
-  targetCastHash: string;
-  schema: 'farcaster.xyz/schemas/v1/cast-remove';
+type CastRemoveBody = {
+  hash: string;
 };
 ```
 
-The remove message must contain the hash of the cast being removed and omit all the other properties. The set can then remove the cast and its contents from the network, which is desirable from a user perspective. The set ensures that the remove message does not reference itself, and that it is only applied as a remove operation if it has a timestamp higher than that of the message it references. The rules for validation are described below, but the rules for merging are specified in each set that implements the remove operation.
+The message must pass the following validation steps: 
 
-<!-- Diagram showing the remove operations -->
+1. `message.data.body.hash` must not equal the `message.envelope.hash`.
```

Comment:
I think just including the rule for `hash` is fine. The others are covered by generic message validation rules.

Reply:
I'm not sure I understand the suggestion

Reply:
The docs include the validation for timestamp and fid in another section. So the only validation that's specific to CastRemove messages is the rule about the hash not referencing itself.

Reply:
ah yeah, i might rework that so leaving this in here for now as i juggle PRs